tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd

```diff
@@ -0,0 +1,584 @@
+
+First we define our dependencies.
+```julia
+using Turing
+using LinearAlgebra
+
+# Packages for visualization
+using VegaLite, DataFrames, StatsPlots
+
+# Import Fisher's iris example data set
+using RDatasets
+
+# Set a seed for reproducibility.
+using Random
+Random.seed!(1789);
+```
+
+Here, we simulate the biological problem described earlier. Admittedly, this is a very simplistic example with far fewer cells and genes and a very straighforward relationship. We try to capture the essence of the problem, but then, unfortunately, real life problems tend to be much more messy.
+```julia
+n_cells = 60
+n_genes = 9
+mu_1 = 10. * ones(n_genes÷3)
+mu_0 = zeros(n_genes÷3)
+S = I(n_genes÷3)
+mvn_0 = MvNormal(mu_0, S)
+mvn_1 = MvNormal(mu_1, S)
+
+# create a diagonal block like expression matrix, with some non-informative genes;
+# not all features/genes are informative, some might just not differ very much between cells)
+expression_matrix = transpose(vcat(
+  hcat(rand(mvn_1, n_cells÷2),
+       rand(mvn_0, n_cells÷2)),
+  hcat(rand(mvn_0, n_cells÷2),
+       rand(mvn_0, n_cells÷2)),
+  hcat(rand(mvn_0, n_cells÷2),
+       rand(mvn_1, n_cells÷2))))
+
+
+df_exp = DataFrame(expression_matrix, :auto)
+df_exp[!,:cell] = 1:n_cells
+
+DataFrames.stack(df_exp, 1:n_genes) |>
+    @vlplot(:rect, x="cell:o", color=:value,
+      encoding={y={field="variable",
+                   type="nominal",
+                   sort="-x",
+                   axis={title="gene"}
+                  }
+               })
+```
+
+Here, you see the simulated data. You can see two groups of cells that differ in the expression of genes. While the difference between the two groups of cells here is fairly obvious from looking at the raw data, in practice and with large enough data sets, it is often impossible to spot the differences from the raw data alone. If you have some patience and compute resources you can increase the size of the dataset, or play around with the noise levels to make the problem increasingly harder.
+
+### pPCA model
+```julia
+@model function pPCA(x, ::Type{TV} = Array{Float64}) where {TV}
+  # Dimensionality of the problem.
+  N, D = size(x)
+
+  # latent variable z
+  z ~ filldist(Normal(), D, N)
+
+  # side note for the curious
+  # we use the more concise filldist syntax partly for compactness, but also for compatibility with other AD
+  # backends, see the [Turing Performance Tipps](https://turing.ml/dev/docs/using-turing/performancetips)
+  # w = TV{2}(undef, D, D)
+  # for d in 1:D
+  #  w[d, :] ~ MvNormal(ones(D))
+  # end
+
+  # weights/loadings W
+  w ~ filldist(Normal(), D, D)
+
+  # mean offset
+  m ~ MvNormal(ones(D))
+  mu = (w * z .+ m)'
```

Comment:
Are you using 2 spaces or 4 spaces for indentation? Here it is 2 and below it is 4. The Julia convention is 4 spaces. 2 would also be okay, but mixing them up doesn't help readability.

Reply:
I have fixed it to 2 whitespaces throughout now.